README.md
# salesforce-component

# DEPRECATED COMPONENT. USE https://github.com/elasticio/salesforce-component-v2 INSTEAD

## Description
[elastic.io](http://www.elastic.io;) iPaaS component that connects to Salesforce API

### Purpose
Salesforce component is designed for Salesforce API integration.

###  Completeness Matrix
![Salesforse-component Completeness Matrix](https://user-images.githubusercontent.com/36419533/75436046-9a5ef880-595c-11ea-838f-32660c119972.png)

[Salesforse-component Completeness Matrix](https://docs.google.com/spreadsheets/d/1_4vvDLdQeXqs3c8OxFYE80CvpeSC8e3Wmwl1dcEGO2Q/edit?usp=sharing)


### API version
The component uses Salesforce - API Version 45.0, except:
- Deprecated Actions and Triggers - API Version 25.0

### Authentication
Authentication occurs via OAuth 2.0.
In the component repository you need to specify OAuth Client credentials as environment variables:
- ```OAUTH_CLIENT_ID``` - your OAuth client key
- ```OAUTH_CLIENT_SECRET``` - your OAuth client secret

## Create new App in Salesforce

In order to make OAuth work, you need a new App in your Salesforce. During app creation process you will be asked to specify
the callback URL, to process OAuth authentication via elastic.io platform your callback URL should be

```https://your-tenant.elastic.io/callback/oauth2```

More information you can find [here](https://help.salesforce.com/apex/HTViewHelpDoc?id=connected_app_create.htm)

## Credentials

During credentials creation you would need to:
- choose ``Environment``
- enter ``Username`` and ``Password`` in a pop-up window after click on ``Authenticate`` button.
- verify and save your new credentials.
### Limitations
According to [Salesforce documentation](https://help.salesforce.com/articleView?id=remoteaccess_request_manage.htm&type=5)

`Each connected app allows five unique approvals per user. When a sixth approval is made, the oldest approval is revoked.`

You can get error `refresh token has been expired` if the same user account was authenticated with same OAuth Application (OAuth Client) more than 4 times. This is feature of the Salesforce platform that automatically invalidates the oldest refresh_token as soon as a number of given refresh tokens for an individual user account exceeds 4.

## Actions
### Query
Executing a SOQL Query that may return many objects. Each resulting object is emitted one-by-one. Use the Salesforce Object Query Language (SOQL) to search your organization’s Salesforce data for specific information. SOQL is similar to the SELECT statement in the widely used Structured Query Language (SQL) but is designed specifically for Salesforce data. This action allows you to interact with your data using SOQL.
Empty object will be returned, if query doesn't find any data.

#### Input fields description
* **Include deleted** - checkbox, if checked - deleted records will be included into the result list.

#### Input fields description
* **Optional batch size** - A positive integer specifying batch size. If no batch size is specified then results of the query will be emitted one-by-one, otherwise, query results will be emitted in an array of maximum batch size.
* **Allow all results to be returned in a set** - checkbox which allows emitting query results in a single array. `Optional batch size` option is ignored in this case.
* **SOQL Query** - Input field where you should type the SOQL query. E.g. `"SELECT ID, Name from Contact where Name like 'John Smi%'"`
* **Max Fetch Count** - limit for a number of messages that can be fetched. 1,000 is the default value when the variable is not set.

### Create Object
Creates a new Selected Object.
Action creates a single object. Input metadata is fetched dynamically from your Salesforce account. Output metadata is the same as input metadata, so you may expect all fields that you mapped as input to be returned as output.

Note:
In case of an **Attachment** object type you should specify `Body` in base64 encoding. `ParentId` is a Salesforce ID of an object (Account, Lead, Contact) which an attachment is going to be attached to.

#### Input fields description
* **Object** - Input field where you should choose the object type, which you want to find. E.g. `Account`
* **Utilize data attachment from previous step (for objects with a binary field)** - a checkbox, if it is checked and an input message contains an attachment and specified object has a binary field (type of base64) then the input data is put into object's binary field. In this case any data specified for the binary field in the data mapper is discarded.

This action will automatically retrieve all existing fields of chosen object type that available on your Salesforce organization

#### Limitations
When **Utilize data attachment from previous step (for objects with a binary field)** is checked and this action is used with Local Agent error would be thrown: 'getaddrinfo ENOTFOUND steward-service.platform.svc.cluster.local steward-service.platform.svc.cluster.local:8200'

### Delete Object (at most 1)
Deletes an object by a selected field. One can filter by either unique fields or all fields of that sobject. Input metadata is fetched dynamically from your Salesforce account.

#### Input field description
* **Object** - dropdown list where you should choose the object type, which you want to find. E.g. `Account`.
* **Type Of Search** -  dropdown list with two values: `Unique Fields` and `All Fields`.
* **Lookup by field** - dropdown list with all fields on the selected object, if on *Type Of Search* is chosen `All Fields`, or with all fields on the selected object where `type` is `id` or `unique` is `true` , if on *Type Of Search* is chosen `Unique Fields` then all searchable fields both custom and standard will be available for selection.

Result is an object with 3 fields.
* **id** - `string`, salesforce object id
* **success** - `boolean`, if operation was successful `true`
* **errors** - `array`, if operation fails, it will contain description of errors

#### Metadata description
Metadata for each particular `Object type` + `Lookup by field` is generating dynamically.

### Upsert Object
Creates or Updates Selected Object.
Action creates a single object. Input metadata is fetched dynamically from your Salesforce account. Output metadata is the same as input metadata, so you may expect all fields that you mapped as input to be returned as output.

#### Input field description
* **Object** - Input field where you should choose the object type, which you want to find. E.g. `Account`
* **Optional Upsert field** - Input field where you should specify the ExternalID name field. E.g. `ExtId__c`.
* **Utilize data attachment from previous step (for objects with a binary field)** - a checkbox, if it is checked and an input message contains an attachment and specified object has a binary field (type of base64) then the input data is put into object's binary field. In this case any data specified for the binary field in the data mapper is discarded.

You should specify **external** or **internal Id** for making some updates in salesforce object.
If you want to create new Object you should always specify **Optional Upsert field** and value of ExternalId in input body structure.

#### Limitations
When **Utilize data attachment from previous step (for objects with a binary field)** is checked and this action is used with Local Agent error would be thrown: 'getaddrinfo ENOTFOUND steward-service.platform.svc.cluster.local steward-service.platform.svc.cluster.local:8200'

### Lookup Object (at most 1)
Lookup an object by a selected field.
Action creates a single object. Input metadata is fetched dynamically from your Salesforce account. Output metadata is the same as input metadata, so you may expect all fields that you mapped as input to be returned as output.

#### Input field description
* **Object** - Dropdown list displaying all searchable object types. Select one type to query, e.g. `Account`.
* **Type Of Search** - Dropdown list with two values: `Unique Fields` and `All Fields`.
* **Lookup by field** - Dropdown list with all fields on the selected object if the *Type Of Search* is `All Fields`. If the *Type Of Search* is `Unique Fields`, the dropdown lists instead all fields on the selected object where `type` is `id` or `unique` is `true`.
* **Include linked objects** - Multiselect dropdown list with all the related child and parent objects of the selected object type. List entries are given as `Object Name/Reference To (Relationship Name)`. Select one or more related objects, which will be join queried and included in the response from your Salesforce Organization. Please see the **Limitations** section below for use case advisories.
* **Allow zero results** - Checkbox. If checked and nothing is found in your Salesforce Organization, an empty object will be returned. If not checked and nothing is found, the action will throw an error.
* **Pass binary data to the next component (if found object has it)** - Checkbox. If it is checked and the found object record has a binary field (primitive type `base64`), then its data will be passed to the next component as a binary attachment.
* **Enable Cache Usage** - Flag to enable cache usage.

#### Metadata description

Metadata contains one field whose name, type and mandatoriness are generated according to the value of the configuration fields *Lookup by field* and *Allow criteria to be omitted*.

#### Limitations
When a binary field (primitive type `base64`, e.g. Documents, Attachments, etc) is selected on **Include linked objects**, an error will be thrown: 'MALFORMED_QUERY: Binary fields cannot be selected in join queries. Instead of querying objects with binary fields as linked objects (such as children Attachments), try querying them directly.' There is also a limit to the number of linked objects that you can query at once - beyond two or three, depending on the number of fields in the linked objects, Salesforce could potentially return a Status Code 431 or 414 error, meaning the query is too long. Finally, due to a bug with multiselect dropdowns, it is recommended to deselect all of the elements in this field before you change your selection in the *Object* dropdown list.

When **Pass binary data to the next component (if found object has it)** is checked and this action is used with Local Agent, an error will be thrown: 'getaddrinfo ENOTFOUND steward-service.platform.svc.cluster.local steward-service.platform.svc.cluster.local:8200'

#### Note
Action has caching mechanism. By default action stores last 10 request-response pairs for 10 min duration.
This parameters can be changed by setting environment variables:
* **HASH_LIMIT_TIME** - Hash expiration time in milis
* **HASH_LIMIT_ELEMENTS** - Hash size number limit

### Lookup Objects
Lookup a list of objects satisfying specified criteria.

#### Input field description
* **Object** - dropdown list where you should choose the object type, which you want to find. E.g. `Account`.
* **Include deleted** - checkbox, if checked - deleted records will be included into the result list.
* **Output method** - dropdown list with following values: "Emit all", "Emit page", "Emit individually".
* **Number of search terms** - text field to specify a number of search terms (positive integer number [1-99] or 0).
* **Enable Cache Usage** - Flag to enable cache usage.
* **Max Fetch Count** - limit for a number of messages that can be fetched. 1,000 is the default value when the variable is not set.

#### Note
Action has caching mechanism. By default action stores last 10 request-response pairs for 10 min duration.
This parameters can be changed by setting environment variables:
* **HASH_LIMIT_TIME** - Hash expiration time in milis
* **HASH_LIMIT_ELEMENTS** - Hash size number limit

#### Metadata description

Depending on the the configuration field *Output method* the input metadata can contain different fields:
*Output method* - "Emit page":
Field "Page size" - optional positive integer that defaults to 1000;
Field "Page number" - required non-negative integer (starts with 0, default value 0);

*Output method* - "Emit all":
Field "Maximum number of records" - optional positive integer (default value 1000);

*Output method* - "Emit individually":
Field "Maximum number of records" - optional positive integer (default value 10000);

Note that the number of records the component emits may affect the performance of the platform/component.

Groups of fields for each search term go next:

Field "Field name" - string represents object's field (a list of allowed values is available);
Field "Field value" - string represents value for selected field;
Field "Condition" - one of the following: "=", "!=", "<", "<=", ">", ">=", "LIKE", "IN", "NOT IN";

Between each two term's group of fields:

Field "Logical operator" - one of the following: "AND", "OR";

Output data is an object, with a field "results" that is an array of objects.

### Bulk Create/Update/Delete/Upsert
Bulk API provides a simple interface for quickly loading large amounts of data from CSV file into Salesforce (up to 10'000 records).
Action takes a CSV file from the attachment as an input. CSV file format is described in the [Salesforce documentatio](https://developer.salesforce.com/docs/atlas.en-us.api_bulk_v2.meta/api_bulk_v2/datafiles.htm)

#### Input field description
* **Operation** - dropdown list with 3 supported operations: `Create`, `Update` and `Delete`.
* **Object** - dropdown list where you should choose the object type to perform bulk operation. E.g. `Case`.
* **Timeout** - maximum time to wait until the server completes a bulk operation (default: `600` sec).

#### Metadata description
* **External ID Field** - a name of the External ID field for `Upsert` operation. E.g. `my_external_id__c`

Result is an object with a property **result**: `array`. It contains objects with 3 fields.
* **id** - `string`, salesforce object id
* **success** - `boolean`, if operation was successful `true`
* **errors** - `array`, if operation failed contains description of errors

#### Limitations
* No errors thrown in case of failed Object Create/Update/Delete/Upsert (`"success": "false"`).
* Object ID is needed for Update and Delete.
* External ID is needed for Upsert.
* Salesforce processes up to 10'000 records from the input CSV file.


### Bulk Query
Fetches records to a CSV file.

#### Input field description
* **SOQL Query** - Input field where you should type the SOQL query. E.g. `"SELECT ID, Name from Contact where Name like 'John Smi%'"`

Result is a CSV file in the attachment.


### Lookup Object (deprecated)
Lookup an object by a selected field.
Action creates a single object. Input metadata is fetched dynamically from your Salesforce account. Output metadata is the same as input metadata, so you may expect all fields that you mapped as input to be returned as output.

#### Input field description
* **Optional batch size** - A positive integer specifying batch size. If no batch size is specified then results of the query will be emitted one-by-one, otherwise, query results will be emitted in an array of maximum batch size.
* **Object** - Input field where you should choose the object type, which you want to find. E.g. `Account`
* **Lookup field** - Input field where you should choose the lookup field which you want to use for result filtering. E.g. `Id`.
* **Max Fetch Count** - limit for a number of messages that can be fetched. 1,000 is the default value when the variable is not set.

```For now, you can specify all unique, lookup, ExternalID/Id fields. ```

##### Execution result handling
|Condition | Execution result |
|----------|------------------|
|Lookup failed - we were not able to find any parent object. |Lookup action emits a single message with an empty body.|
|Lookup found a single object, e.g. we were able to identify a parent Account to the Contact|A single message will be emitted, found object will be a body of the message|
|Lookup found multiple objects (that may happen when a lookup is made by non-unique field) | Each found object will be emitted with the separate message|

### New Account `(deprecated)`
Creates a new Account.
Action creates a single object. Input metadata is fetched dynamically from your Salesforce account. Output metadata is the same as input metadata, so you may expect all fields that you mapped as input to be returned as output.

#### Input fields description
This action will automatically retrieve all existing fields of `Account` object type that available on your Salesforce organization.

Action is `deprecated`. You can use [Create Object](#create-object) action instead.

### New Case `(deprecated)`
Creates a new Case.
Action creates a single object. Input metadata is fetched dynamically from your Salesforce account. Output metadata is the same as input metadata, so you may expect all fields that you mapped as input to be returned as output.

#### Input fields description
This action will automatically retrieve all existing fields of `Case` object type that available on your Salesforce organization

Action is `deprecated`. You can use [Create Object](#create-object) action instead.

### New Contact `(deprecated)`
Creates a new Contact.
Action creates a single object. Input metadata is fetched dynamically from your Salesforce account. Output metadata is the same as input metadata, so you may expect all fields that you mapped as input to be returned as output.


#### Input fields description
This action will automatically retrieve all existing fields of `Contact` object type that available on your Salesforce organization

Action is `deprecated`. You can use [Create Object](#create-object) action instead.

### New Event `(deprecated)`
Creates a new Event.
Action creates a single object. Input metadata is fetched dynamically from your Salesforce account. Output metadata is the same as input metadata, so you may expect all fields that you mapped as input to be returned as output.

#### Input fields description
This action will automatically retrieve all existing fields of `Event` object type that available on your Salesforce organization

Action is `deprecated`. You can use [Create Object](#create-object) action instead.

### New Lead `(deprecated)`
Creates a new Lead.
Action creates a single object. Input metadata is fetched dynamically from your Salesforce account. Output metadata is the same as input metadata, so you may expect all fields that you mapped as input to be returned as output.

#### Input fields description
This action will automatically retrieve all existing fields of `Lead` object type that available on your Salesforce organization

Action is `deprecated`. You can use [Create Object](#create-object) action instead.

### New Note `(deprecated)`
Creates a new Note.
Action creates a single object. Input metadata is fetched dynamically from your Salesforce account. Output metadata is the same as input metadata, so you may expect all fields that you mapped as input to be returned as output.

#### Input fields description
This action will automatically retrieve all existing fields of `Note` object type that available on your Salesforce organization

Action is `deprecated`. You can use [Create Object](#create-object) action instead.

### New Task `(deprecated)`
Creates a new Task.
Action creates a single object. Input metadata is fetched dynamically from your Salesforce account. Output metadata is the same as input metadata, so you may expect all fields that you mapped as input to be returned as output.

#### Input fields description
This action will automatically retrieve all existing fields of `Task` object type that available on your Salesforce organization

Action is `deprecated`. You can use [Create Object](#create-object) action instead.

## Triggers
### Query
Continuously runs the same SOQL Query and emits results one-by-one.
Use the Salesforce Object Query Language (SOQL) to search your organization’s Salesforce data for specific information. SOQL is similar to the SELECT statement in the widely used Structured Query Language (SQL) but is designed specifically for Salesforce data. This action allows you to interact with your data using SOQL.

#### List of Expected Config fields

* **SOQL Query** - Input field for your SOQL Query
* **Output method** - dropdown list with options: `Emit all` - all found records will be emitted in one array `records`, and `Emit individually` - each found object will be emitted individual. Optional field, defaults to: `Emit individually`.

NOTE: Max possible fetch size is 2000 objects per execution.

### Get New and Updated Objects Polling

CHANGELOG.md
## 1.3.10 (December 09, 2022)

* Update Sailor version to 2.7.1

## 1.3.9 (April 08, 2022)

* Update Sailor version to 2.6.27
* Get rid of vulnerabilities in dependencies
* Add component pusher job to Circle.ci config

## 1.3.8 (November 10, 2020)

* Change logo to grayscale.

## 1.3.7 (October 30, 2020)

* Upgrade to sailor 2.6.18

## 1.3.6 (October 8, 2020)

* Component got deprecated. Use https://github.com/elasticio/salesforce-component-v2 instead

## 1.3.5 (August 21, 2020)

* Update `Bulk Create/Update/Delete` action:
   - now it supports `Bulk Upsert` feature
   - fix bug `404 - File's metadata is not found`

## 1.3.4 (May 8, 2020)

* Hotfix: removed `Max Fetch Count` field from Query trigger, as it does not work

## 1.3.3 (May 8, 2020)

* Fix bug with 1,000 objects limit in actions:
  - Query
  - Lookup Object
  - Lookup Objects

  and trigger:
  - Get New and Updated Objects Polling.

New configuration field `Max Fetch Count` added to configure the limit. Please see the README.md for the details

## 1.3.2 (March 26, 2020)

* Add new optional field `Include linked objects` in trigger: `Get New and Updated Objects Polling`

## 1.3.1 (March 11, 2020)

* Add new optional field `Output method` in triggers: `Query` and `Get New and Updated Objects Polling`

## 1.3.0 (February 27, 2020)

* Add Delete Object (at most 1) Action
* Add new optional field in the Lookup Object Action

## 1.2.3 (February 4, 2020)

* Fix query action

## 1.2.2 (January 25, 2020)

* Add request caching for lookup actions

## 1.2.1 (December 27, 2019)

* Update sailor version to 2.5.4
* Refactor console.log to built in sailor logger
* Change build type to `docker`

## 1.2.0 (December 2, 2019)

* Add support for `Bulk operations` feature (Create/Update/Delete and Query)
* Add `Delete Object` action
* Add `Lookup Objects` action
* `Create object` action: add ability to utilize binary data attachment from previous step
* `Upsert object` action: add ability to utilize binary data attachment from previous step
* `Lookup Object (at most 1)` action: add ability to pass binary data (if found object has it) to the next component as a binary attachment
* `Query` action: add ability to query deleted objects

## 1.1.2 (October 28, 2019)

* Change Oauth values naming

## 1.1.1 (July 10, 2019)

* Add support for `Create Attachment` feature
* Fix bug with Salesforce's and platform's types mismatch
* Make unit tests great again (internal issue)

## 1.0.0 (June 27, 2019)

* Initial release which includes a bunch of previous unversioned releases


lib/entry.js
/* eslint-disable no-use-before-define,no-param-reassign,no-await-in-loop */
const _ = require('lodash');
const util = require('util');
const Q = require('q');
const jsforce = require('jsforce');
const elasticio = require('elasticio-node');

const { messages } = elasticio;
const httpUtils = require('./helpers/http-utils');
const converter = require('./helpers/metadata');
const describe = require('./helpers/describe');
const MetaLoader = require('./helpers/metaLoader');
const fetchObjectsQuery = require('./helpers/objectFetcherQuery');
const createPresentableError = require('./helpers/error.js');
const oAuthUtils = require('./helpers/oauth-utils.js');
const common = require('./common.js');

exports.SalesforceEntity = SalesforceEntity;

function SalesforceEntity(callScope) {
  const self = this;

  /**
   * This function refreshes salesforce token
   * @param conf - configuration with so that conf.oauth.refresh_token should be available
   * @param next - callback that will be called with (err, conf) parameters
   */
  this.refreshToken = function refreshToken(conf, next) {
    oAuthUtils.refreshAppToken(callScope.logger, 'salesforce', conf,
      // eslint-disable-next-line consistent-return
      (err, newConf) => {
        if (err) {
          return next(err);
        }
        callScope.emit('updateKeys', { oauth: newConf.oauth });
        next(null, newConf);
      });
  };

  this.getInMetaModel = function getInMetaModel(cfg, cb) {
    getMetaModel('in', cfg, cb);
  };

  this.getOutMetaModel = function getOutMetaModel(cfg, cb) {
    getMetaModel('out', cfg, cb);
  };

  function getMetaModel(direction, cfg, cb) {
    // eslint-disable-next-line consistent-return
    self.refreshToken(cfg, (err, newCfg) => {
      if (err) {
        callScope.logger.error('Error refreshing token');
        return cb(err);
      }
      describe.getObjectDescription(callScope.logger, newCfg,
        (errInner, objectDescription) => {
          if (errInner) {
            cb(errInner);
          } else {
            const metadata = {};
            metadata[direction] = converter.buildSchemaFromDescription(
              objectDescription, direction,
            );
            cb(null, metadata);
          }
        });
    });
  }

  /**
   * Polling function that will pull changes from Salesforce
   *
   * @param msg - message is null
   * @param cfg - configuration
   * @param snapshot
   */
  this.processTrigger = async function processTrigger(msg, cfg, snapshot) {
    const {
      linkedObjects = [],
    } = cfg;

    if (!snapshot || (typeof (snapshot) === 'object' && !snapshot.previousLastModified)) {
      snapshot = {
        previousLastModified: cfg.startTime || '1970-01-01T00:00:00.000Z',
      };
    } else if (typeof snapshot === 'string') {
      // for backward compatibility
      snapshot = { previousLastModified: snapshot };
    }

    const conn = new jsforce.Connection({
      oauth2: {
        clientId: process.env.OAUTH_CLIENT_ID,
        clientSecret: process.env.OAUTH_CLIENT_SECRET,
      },
      instanceUrl: cfg.oauth.instance_url,
      accessToken: cfg.oauth.access_token,
      refreshToken: cfg.oauth.refresh_token,
      version: common.globalConsts.SALESFORCE_API_VERSION,
    });
    conn.on('refresh', (accessToken, res) => {
      callScope.logger.debug('Keys were updated');
      emitKeys({ oauth: res });
    });

    let maxFetch = cfg.maxFetch ? parseInt(cfg.maxFetch, 10) : 1000;

    if (cfg.sizeOfPollingPage) {
      callScope.logger.debug('Current sizeOfPollingPage=%s, maxFetch=%s', cfg.sizeOfPollingPage, maxFetch);
      const sizeOfPollingPage = parseInt(cfg.sizeOfPollingPage, 10);
      if (sizeOfPollingPage && sizeOfPollingPage > 0 && sizeOfPollingPage <= 10000) {
        maxFetch = sizeOfPollingPage;
      } else {
        emitError('Size of Polling Page needs to be positive integer, max 10000 objects');
        emitEnd();
        return;
      }
    }
    const singlePagePerInterval = (cfg.singlePagePerInterval !== 'no');
    const maxTime = cfg.endTime ? ` AND LastModifiedDate <= ${cfg.endTime}` : '';
    let hasMorePages = true;
    let lastSeenTime = snapshot.previousLastModified;

    // the query for the object and all its linked parent objects
    const selectedObjects = linkedObjects.reduce((query, obj) => {
      if (!obj.startsWith('!')) return `${query}, ${obj}.*`;
      return query;
    }, '*');

    // selecting the object and all its parents
    let query = conn.sobject(cfg.object)
      .select(selectedObjects);

    // the query for all the linked child objects
    query = linkedObjects.reduce((newQuery, obj) => {
      if (obj.startsWith('!')) {
        return newQuery.include(obj.slice(1))
          .select('*')
          .end();
      }
      return newQuery;
    }, query);

    do {
      let whereCondition;
      if (lastSeenTime === cfg.startTime || lastSeenTime === '1970-01-01T00:00:00.000Z') whereCondition = `LastModifiedDate >= ${lastSeenTime}${maxTime}`;
      else whereCondition = `LastModifiedDate > ${lastSeenTime}${maxTime}`;
      try {
        const results = await query.where(whereCondition)
          .sort({ LastModifiedDate: 1 })
          .execute({ autoFetch: true, maxFetch });
        processResults(results);
      } catch (err) {
        emitError(createPresentableError(err) || err);
        emitEnd();
        return;
      }
      if (singlePagePerInterval) {
        hasMorePages = false;
      }
    } while (hasMorePages);

    if (snapshot.previousLastModified !== lastSeenTime) {
      snapshot.previousLastModified = lastSeenTime;
      callScope.logger.debug('emitting new snapshot');
      emitSnapshot(snapshot);
    }
    emitEnd();

    function processResults(records) {
      if (!records || !records.length) {
        callScope.logger.info('No new objects found');
        hasMorePages = false;
        return;
      }
      const { outputMethod = 'emitIndividually' } = cfg;
      if (outputMethod === 'emitAll') {
        emitData(messages.newMessageWithBody({ records }));
      } else if (outputMethod === 'emitIndividually') {
        records.forEach((record, i) => {
          const newMsg = messages.newEmptyMessage();
          newMsg.headers = {
            objectId: record.attributes.url,
          };
          newMsg.body = record;
          callScope.logger.info('emitting record %d', i);
          emitData(newMsg);
        });
      } else {
        throw new Error('Unsupported Output method');
      }
      hasMorePages = records.length === maxFetch;
      lastSeenTime = records[records.length - 1].LastModifiedDate;
    }
  };

  this.processQuery = function processQuery(query, cfg) {
    const params = {};
    params.cfg = cfg;
    params.cfg.apiVersion = `v${common.globalConsts.SALESFORCE_API_VERSION}`;

    params.query = query;

    Q.ninvoke(self, 'refreshToken', params.cfg)
      .then(updateCfg)
      .then(paramsUpdated => fetchObjectsQuery(callScope.logger, paramsUpdated))
      .then(processResults)
      .fail(onError)
      .done(emitEnd);

    function updateCfg(config) {
      params.cfg = config;
      return params;
    }

    function processResults(results) {
      if (!results.objects.totalSize) {
        callScope.logger.info('No new objects found');
        return;
      }
      const { records } = results.objects;
      const { outputMethod = 'emitIndividually' } = params.cfg;
      if (outputMethod === 'emitIndividually') {
        records.forEach(emitResultObject);
      } else if (outputMethod === 'emitAll') {
        emitData(messages.newMessageWithBody({ records }));
      } else {
        throw new Error('Unsupported Output method');
      }
    }

    function emitResultObject(object) {
      const msg = messages.newEmptyMessage();
      msg.headers = {
        objectId: object.attributes.url,
      };
      msg.body = object;
      emitData(msg);
    }
  };

  /**
   * This function will fetch object data from salesforce
   * @param conf
   * @param next
   */
  this.listObjectTypes = function listObjectTypes(conf, next) {
    // eslint-disable-next-line consistent-return
    self.refreshToken(_.clone(conf), (err, newCfg) => {
      if (err) {
        return next(err);
      }
      describe.fetchObjectTypes(callScope.logger, newCfg, next);
    });
  };

  this.listLinkedObjects = function listLinkedObjects(conf, next) {
    // eslint-disable-next-line consistent-return
    self.refreshToken(_.clone(conf), async (err, newCfg) => {
      if (err) {
        return next(err);
      }
      try {
        newCfg.sobject = newCfg.object;
        const metaLoader = new MetaLoader(newCfg, this);
        const relatedObjs = await metaLoader.getLinkedObjectsModel();
        next(null, relatedObjs);
      } catch (errInner) {
        next(errInner);
      }
    });
  };

  /**
   * This function modify a scope and add two functions process and getMetaModel
   *
   * @param objectType
   * @param msg
   * @param conf
   */
  this.processAction = function processAction(objectType, msg, conf) {
    // eslint-disable-next-line consistent-return
    self.refreshToken(conf, (err, newCfg) => {
      if (err) {
        return onActionError(err);
      }
      postData(newCfg, checkPostResult, common.globalConsts.SALESFORCE_API_VERSION);
    });

    function postData(config, next) {
      // Here we assume conf has a refreshed Salesforce token
      const baseUrl = config.oauth.instance_url;
      const authValue = util.format('Bearer %s', config.oauth.access_token);
      const url = util.format('%s/services/data/%s/sobjects/%s', baseUrl, `v${common.globalConsts.SALESFORCE_API_VERSION}`,
        objectType);

      callScope.logger.debug('getJSON');

      httpUtils.getJSON(callScope.logger, {
        url,
        method: 'POST',
        json: msg.body,
        auth: authValue,
        statusExpected: 201,
      }, next);
    }

    // eslint-disable-next-line consistent-return
    function checkPostResult(err, result) {
      if (err) {
        return onActionError(err);
      }
      msg.body = _.extend(msg.body, result);
      onActionSuccess(msg);
    }

    function onActionError(err) {
      emitError(err);
      emitEnd();
    }

    function onActionSuccess(message) {
      emitData(message);
      emitEnd();
    }
  };

  function emitError(err) {
    callScope.logger.error('emitting SalesforceEntity error');
    callScope.emit('error', err);
  }

  function onError(err) {
    emitError(createPresentableError(err) || err);
  }

  function emitData(data) {
    callScope.emit('data', data);
  }

  function emitSnapshot(snapshot) {
    callScope.emit('snapshot', snapshot);
  }

  function emitKeys(snapshot) {
    callScope.emit('updateKeys', snapshot);
  }

  function emitEnd() {
    callScope.emit('end');
  }
}

exports.SalesforceEntity = SalesforceEntity;

exports.buildAction = function buildAction(objectType, exports) {
  exports.process = function processAction(msg, conf) {
    const self = new SalesforceEntity(this);
    self.processAction(objectType, msg, conf, common.globalConsts.SALESFORCE_API_VERSION);
  };

  exports.getMetaModel = function getActionMetaModel(cfg, cb) {
    const self = new SalesforceEntity(this);
    cfg.object = objectType;
    cfg.sobject = cfg.sobject ? cfg.sobject : objectType;
    // eslint-disable-next-line consistent-return
    self.getInMetaModel(cfg, (err, data) => {
      if (err) {
        return cb(err);
      }
      data.out = _.cloneDeep(data.in);
      data.out.properties.in = {
        type: 'string',
        required: true,
      };
      cb(null, data);
    });
  };
};

exports.buildTrigger = function buildTrigger(objectType, exports) {
  exports.process = function processTrigger(msg, conf, snapshot) {
    const self = new SalesforceEntity(this);
    // Set fixed object type to the configuration object
    conf.object = objectType;
    // Proceed as usual
    self.processTrigger(msg, conf, snapshot);
  };

  exports.getMetaModel = function getTriggerMetaModel(cfg, cb) {
    const self = new SalesforceEntity(this);
    cfg.object = objectType;
    self.getOutMetaModel(cfg, cb);
  };
};

exports.objectTypes = function objectTypes(conf, next) {
  const self = new SalesforceEntity(this);
  self.listObjectTypes(conf, next);
};

exports.linkedObjectTypes = function linkedObjectTypes(conf, next) {
  const self = new SalesforceEntity(this);
  self.listLinkedObjects(conf, next);
};

exports.process = function processEntry(msg, conf, snapshot) {
  const self = new SalesforceEntity(this);
  self.processTrigger(msg, conf, snapshot);
};

exports.getMetaModel = function getEntryMetaModel(cfg, cb) {
  const self = new SalesforceEntity(this);
  self.getOutMetaModel(cfg, cb);
};
